File: Algorithm/BOJ/java/1725.java
//https://www.acmicpc.net/problem/1725

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

import static java.lang.Math.max;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        Stack<Integer> stack = new Stack<>();
        int n = Integer.parseInt(br.readLine());
        int[] histo = new int[n + 1];
        int ans = 0;
        int top;

        for(int i = 0; i <= n; i++) {
            if(i < n) histo[i] = Integer.parseInt(br.readLine());
            while(!stack.empty() && histo[stack.peek()] > histo[i]) {
                int bot;
                top = stack.peek();
                stack.pop();
                if(stack.empty()) bot = i;
                else bot = i - stack.peek() - 1;

                ans = max(ans, histo[top] * bot);
            }
            stack.push(i);
        }
        System.out.println(ans);
    }
}

File: Algorithm/BOJ/cpp/1725.cpp
//https://www.acmicpc.net/problem/1725
//http://1ilsang.blog.me/221412218805

#include <stack>
#include <cstdio>
#include <algorithm>

using namespace std;

int main(void) {
    stack<int> st;
    int ans = 0;
    int n, top;
    int histo[100001];

    scanf("%d", &n);
    for (int i = 0; i < n; i++) scanf("%d", &histo[i]);

    for (int i = 0; i <= n; i++) {
        while (!st.empty() && histo[st.top()] >= histo[i]) {
            int bot;
            top = st.top();
            st.pop();
            if (st.empty()) bot = i;
            else bot = i - st.top() - 1;

            ans = max(ans, histo[top] * bot);
        }
        st.push(i);
    }

    printf("%d\n", ans);

    return 0;
}
